### PR TITLE
fix demo in epiphany

### LINKED ISSUE #4223: demo is broken with Epiphany browser
Currently the demo will not start up with the Epiphany browser (webkit based, gnome's default browser), because of the webgl addon being set as default.

Prolly linked to #3357, I think we need to find a more general conditional than `isSafari` for those webkit driven browsers, along with a more fine-grained version inspection (assuming newer webkit versions work already?)

### PR DESCRIPTION
**Pull request by jerch**
Fixes #4223. (also prolly for older safari/webkit in general)

Note that beside the addon loading logic, there was also an issue in the canvas renderer  - for some reason some glyphs have no width/height, which older safari/webkit does not like. Not sure if that needs further investigation/fix on why glyphs might be empty.

### CODE REVIEW

**Comment by Tyriar** [comment]
Does that browser have fingerprint detection prevention? There's issue in Brave because of this where it blocks access to various canvas font APIs without actually telling the web app.

File: demo/client.ts [hunk 1]
```diff
@@ -273,22 +273,22 @@ function createTerminal(): void {
...
   addons.fit.instance!.fit();
-  typedTerm.loadAddon(addons.webgl.instance);
-  setTimeout(() => {
-    if (addons.webgl.instance !== undefined) {
-      setTextureAtlas(addons.webgl.instance.textureAtlas);
-      addons.webgl.instance.onChangeTextureAtlas(e => setTextureAtlas(e));
-      addons.webgl.instance.onAddTextureAtlasCanvas(e => appendTextureAtlas(e));
-      addons.webgl.instance.onRemoveTextureAtlasCanvas(e => removeTextureAtlas(e));
-    }
-  }, 0);
```
[the remark this answers is not in the document]
Comment (jerch) [reply demo/client.ts:284]: Well it works in all browsers that support webgl - the texture atlas page is correctly shown below after a first load. Guess it gets spawned with sync code during `term.open` call? The try catch cannot move as a whole into setTimeout, as `loadAddon` already throws for older safari versions.

We could move this safari version check: https://github.com/xtermjs/xterm.js/blob/497df06f3f848ce1ae2b33dd69b2e56a2791ff3e/addons/xterm-addon-webgl/src/WebglAddon.ts#L36-38
into WebglAddon ctor to spot engine incompatibility earlier during addon instantiation (seems more logical to me to refuse instantiation at all, instead of having a dysfunctional instance floating around)?

File: addons/xterm-addon-canvas/src/BaseRenderLayer.ts [hunk 2]
```diff
@@ -373,6 +373,9 @@ export abstract class BaseRenderLayer extends Disposable implements IRenderLayer
     } else {
       glyph = this._charAtlas.getRasterizedGlyph(cell.getCode() || WHITESPACE_CELL_CODE, this._cellColorResolver.result.bg, this._cellColorResolver.result.fg, this._cellColorResolver.result.ext);
     }
+    if (!glyph.size.x || !glyph.size.y) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (jerch) [reply addons/xterm-addon-canvas/src/BaseRenderLayer.ts:378]: Re-checked this - its caused by NULL char and SP, which I guess has no glyph repr at all (its just whitespace in the end). I suggest an earlier filtering of those in the `TextRenderer` to avoid bad runtime. Currently it runs into this edge case for all empty cells:

Tested with `ls -lR /usr`:
- current: ~3800 msec script runtime
- with early empty filtering: ~3000 msec script runtime

While testing it, I found another issue - the rect-clipping done in that method is very expensive and leads to many unfinished render frames in my profiling data (~every second frame gets not finished). Commenting the rect-clipping out changes that dramatically, where almost all frames get rendered in time. Runtime differences:
- with early empty filtering: ~3000 msec script runtime
- with early empty filtering and no clipping: ~2400 msec script runtime

So for the rect-clipping I suggest to rethink the way it is currently done, maybe clip the glyphs during creation already?

In summary both issues create ~50% worse runtime for canvas rendering. Also note that the difference is perceivable since the new release (already wondered about it).

[the remark this answers is not in the document]
Comment (Tyriar) [reply addons/xterm-addon-canvas/src/BaseRenderLayer.ts:378]: Ah ok that makes sense, empty glyphs use the null glyph here:

https://github.com/jerch/xterm.js/blob/309ba47483840a616931122877c4e5d93d56aab9/src/browser/renderer/shared/TextureAtlas.ts#L22-L29

In webgl this is pointed at which ends up printing nothing, seems like a decent solution for canvas here but it could do with a comment.

> While testing it, I found another issue - the rect-clipping done in that method is very expensive and leads to many unfinished render frames in my profiling data (~every second frame gets not finished).

Do you mean `_clipRow`? This is pretty essential to allow rendering partial viewport updated, preventing artifacts going on other lines.

[the remark this answers is not in the document]
Comment (Tyriar) [reply addons/xterm-addon-canvas/src/BaseRenderLayer.ts:378]: > Yes, _clipRow introduces really bad runtime penalty.

I think the row clipping is just the way you're meant to do this in canvas, so if it's slow I doubt we'd be able to make it faster by taking a more manual approach. Since it's the canvas renderer I'm not too concerned if it's not highly optimized as it's basically just the fallback for older versions of webkit which haven't got the webgl2 support yet.

> So what would be the right fix here?

What you have seems fine, you can add an early exit for the space character but it's overcomplicating things to try exit early for any other characters imo.

[the remark this answers is not in the document]
Comment (jerch) [reply addons/xterm-addon-canvas/src/BaseRenderLayer.ts:378]: > Since it's the canvas renderer I'm not too concerned if it's not highly optimized as it's basically just the fallback for older versions of webkit ...

But look, all those sad pandas with older ipads :open_mouth: 

> What you have seems fine, you can add an early exit for the space character but it's overcomplicating things to try exit early for any other characters imo.

Yeah will add an early skip clause for that, as it seems to give some better runtime.

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: demo/client.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerch) [reply demo/client.ts]: Changed it.

**Review by Tyriar — APPROVED** [review]
